pilot/pkg/serviceregistry/kube/controller/network.go

```diff
@@ -242,25 +244,65 @@ func (c *Controller) updateServiceNodePortAddresses(svcs ...*model.Service) bool
 		return false
 	}
 	for _, svc := range svcs {
+		c.RLock()
+		_, found := c.nodeSelectorsForServices[svc.Hostname]
+		isNodePortGateway := found && c.servicesMap[svc.Hostname] != nil
+		c.RUnlock()
+
+		if isNodePortGateway && svc.Attributes.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+			endpoints := c.buildEndpointsForService(svc)
```

Comment:
Should `buildEndpointsForService` just do this filtering instead? I'm a little concerned that other code paths might miss this logic. Maybe it's not an issue?

Reply:
I didn't get it. Sorry. There, endpoint computation is done to update `nodePortGwLocalServiceToWorkloadNodesMap` which tracks the nodes with at least one endpoint for a given Service (keyed by hostname). Ideally `buildEndpointsForService` should not be concerned with Node addresses which could be different from Workload addresses (in Kubernetes, it is node IP and pod IPs would be different).

In that particular `if` block, list of node addresses are being updated. For this particular case, it is the subset of the cluster nodes. When external traffic policy is `Cluster` it would be all nodes of the cluster.